### PR TITLE
Feature equip item

### PR DESCRIPTION
**Pull request by PaulQbFeng**
My proposal for item handling 

- Change champion stats structure. All stats have same name defined in glossary.
```
self.orig_base_stats --> for basic champion stats
self.item_stats --> for item stats
self.orig_bonus_stat --> for bonus stats (item + runes)
```
- Can init a champion with a list of items defined in ALL_ITEM_STATS
- can equip dynamically an item with `equip_item`
- I tested ahri lvl 4 with cloth armor, pick axe, long sword, bf autoattack a dummy (see tests/test_champion.py)

I believe we can improve this solution by creating a Stat class and creating a magic method __add__ so we could add stat class between them like `self.bonus_stats = self.item_stats + self.rune_stats`

Still need to test auto attack with a setup with armor pen, letha etc

### CODE REVIEW

**Comment by PaulQbFeng** [comment]
bonus_stats should have been named orig_bonus_stast such as 

all champions have in theory
- orig_base_stats with champion base stats
- orig_bonus_stats computed from items + runes + others
- base_stats that can evolve during a fight 
- bonus_stats that can evolve during a fight

the last 2 can be altered when running `auto_attack` or `spell_Q` (yet to be created)

BaseChampion should have a method `reinitialize_stats /bonus stats`

File: champion.py [hunk 1]
```diff
@@ -1,62 +1,93 @@
...
-        - Equip item
+    Base class to represent a champion. It is initialized with the stats of a champion at a given level.
+    Some mechanisms are shared accross all champions:
+        - equip_item
         - auto attack
-        ...
     """
 
-    def __init__(self, champion_name: str, level: int = 1):
+    def __init__(self, champion_name: str, level: int = 1, items: List[str] = None):
```
Comment (michaelbenayoun) [inline champion.py:17]: `items` should be `Optional[List[str]]`.

File: champion.py [hunk 2]
```diff
@@ -1,62 +1,93 @@
...
+            return dict()
+
+        total_item_stats = dict()
+        for item_name in items:
+            item_stats =  ALL_ITEM_STATS[item_name]
+            self.update_bonus_stat_with_item(total_item_stats, item_stats)
+
+        return total_item_stats
+
+    def update_bonus_stat_with_item(self, total_item_stats, item_stats):
```
Comment (michaelbenayoun) [inline champion.py:42]: You could use a `defaultdict` to not have to distinguish the tha cases.

File: champion.py [hunk 3]
```diff
@@ -1,62 +1,93 @@
...
+            if stat_name not in total_item_stats:
+                total_item_stats[stat_name] = stat_value
+            else:
+                total_item_stats[stat_name] += stat_value
+
+    def get_bonus_stats(self): # TODO: add runes
+        """Get bonus stats from all sources of bonus stats (items, runes)"""
+        if self.items is None:
+            return dict()
+        return self.item_stats.copy()
```
Comment (michaelbenayoun) [inline champion.py:54]: Is it a deepcopy ?

File: champion.py [hunk 4]
```diff
@@ -1,62 +1,93 @@
...
+
+    def equip_item(self, item_name):
+        self.items.append(item_name)
+        self.update_bonus_stat_with_item(self.item_stats, ALL_ITEM_STATS[item_name])
+        self.orig_bonus_stats = self.get_bonus_stats()
 
     def auto_attack(self, enemy_champion):
         """Calculates the damage dealt to an enemy champion with an autoattack"""
-        return damage_after_positive_resistance(self.attack_damage, enemy_champion.armor)
+        bonus_attack_damage = self.orig_bonus_stats["attack_damage"] if "attack_damage" in self.orig_bonus_stats else 0
```
Comment (michaelbenayoun) [inline champion.py:63]: `self.orig_bonus_stats.get("self.orig_bonus_stats", 0)`

File: champion.py [hunk 5]
```diff
@@ -1,62 +1,93 @@
...
+        bonus_armor = enemy_champion.orig_bonus_stats["armor"] if "armor" in enemy_champion.orig_bonus_stats else 0
```
Comment (michaelbenayoun) [inline champion.py:64]: Same comment.

**Review by michaelbenayoun — COMMENTED** [review]
Left a few nits and comments

File: champion.py (hunk 3 above)
Comment (PaulQbFeng) [reply champion.py:54]: the copy was not necessary, fixed

**Pull request merged**